src/plugin/systems/multiple_rapier_contexts.rs

```diff
@@ -0,0 +1,221 @@
+                    || x.entity2body.contains_key(&entity)
+            })
+            .unwrap_or(false)
+        {
+            remove_old_physics(entity, &mut commands);
+
+            bubble_down_world_change(
+                &mut commands,
+                entity,
+                &q_children,
+                *new_physics_world,
+                &q_physics_world,
+            );
+        }
+    }
+}
+
+fn bubble_down_world_change(
+    commands: &mut Commands,
+    entity: Entity,
+    q_children: &Query<&Children>,
+    new_physics_world: RapierContextEntityLink,
+    q_physics_world: &Query<&RapierContextEntityLink>,
+) {
+    let Ok(children) = q_children.get(entity) else {
+        return;
+    };
+
+    children.iter().for_each(|&child| {
+        if q_physics_world
+            .get(child)
+            .map(|x| *x == new_physics_world)
+            .unwrap_or(false)
+        {
+            return;
+        }
+
+        remove_old_physics(child, commands);
+        commands.entity(child).insert(new_physics_world);
+
+        bubble_down_world_change(
+            commands,
+            child,
+            q_children,
+            new_physics_world,
+            q_physics_world,
+        );
+    });
+}
+
+#[cfg(test)]
+mod test {
+    use crate::plugin::systems::tests::HeadlessRenderPlugin;
+    use crate::plugin::{
+        NoUserData, PhysicsSet, RapierContext, RapierContextEntityLink, RapierPhysicsPlugin,
+    };
+    use crate::prelude::{ActiveEvents, Collider, ContactForceEventThreshold, RigidBody, Sensor};
+    use bevy::prelude::*;
+    use bevy::time::{TimePlugin, TimeUpdateStrategy};
+    use rapier::math::Real;
+
+    #[test]
+    pub fn multi_world_hierarchy_update() {
+        let mut app = App::new();
+        app.add_plugins((
+            HeadlessRenderPlugin,
+            TransformPlugin,
+            TimePlugin,
+            RapierPhysicsPlugin::<NoUserData>::default(),
+        ))
+        .add_systems(
+            PostUpdate,
+            setup_physics
+                .run_if(run_once())
+                .before(PhysicsSet::SyncBackend),
+        );
+        // Simulates 60 updates per seconds
+        app.insert_resource(TimeUpdateStrategy::ManualDuration(
+            std::time::Duration::from_secs_f32(1f32 / 60f32),
+        ));
+        app.update();
+        // Verify all rapier entities have a `RapierContextEntityLink`.
+        let world = app.world_mut();
+        let mut query = world.query_filtered::<Entity, With<Marker<'R'>>>();
+        for entity in query.iter(&world) {
+            world
+                .get::<RapierContextEntityLink>(entity)
+                .unwrap_or_else(|| panic!("no link to rapier context entity from {entity}."));
+        }
+        // Verify link is correctly updated for children.
+        let new_rapier_context = world.spawn(RapierContext::default()).id();
+        // FIXME: We need to wait 1 frame when creating a world.
```

Comment:
Is this still a problem?

Reply:
I... believe so. I'll try to track down why exactly.

Reply:
I'd love a test for that specifically ; I'd love to update this test to be able to test exactly an updated world in the same frame, but I'm not sure on where to insert the user systems exactly for that. `Last` seems wrong.